applications/StructuralMechanicsApplication/custom_processes/spr_error_process.cpp

```diff
@@ -0,0 +1,531 @@
+    BoundedMatrix<double, 1, SigmaSize * (TDim + 1)> A2;
+    for (IndexType j = 0; j < SigmaSize; ++j){
+        p_k(j,j * (TDim + 1) + 1)= itNode->X() - itPatchNode->X();
+        p_k(j,j * (TDim + 1) + 2)= itNode->Y() - itPatchNode->Y();
+        if(TDim == 3)
+            p_k(j,j * (TDim + 1) + 3)= itNode->Z() - itPatchNode->Z();
+    }
+    
+    // Set the normal and tangential vectors in Voigt Notation
+    const array_1d<double, 3>& normal = itNode->GetValue(NORMAL);
+
+    if(TDim == 2) {
+        N_k(0,0) = normal[0] * normal[0];
+        N_k(0,1) = normal[1] * normal[1];
+        N_k(0,2) = 2.0 * normal[0] * normal[1];
+        
+        T_k1(0,0) =  normal[0] * normal[1];
+        T_k1(0,1) = -normal[0] * normal[1];
+        T_k1(0,2) =  normal[1] * normal[1] - normal[0] * normal[0];
+    } else if (TDim == 3) {
+        N_k(0,0) = normal[0] * normal[0];
+        N_k(0,1) = normal[1] * normal[1];
+        N_k(0,1) = normal[2] * normal[2];
+        N_k(0,3) = 2.0 * normal[1] * normal[2];
+        N_k(0,4) = 2.0 * normal[2] * normal[0];
+        N_k(0,5) = 2.0 * normal[0] * normal[1];
+
+        // Set tangential vectors
+        array_1d<double, 3> t1(3, 0.0);
+        array_1d<double, 3> t2(3, 0.0);
+
+        const double tolerance = std::numeric_limits<double>::epsilon();
+        if(std::abs(normal[0]) > tolerance || std::abs(normal[1]) > tolerance) {
+            const double norm = std::sqrt((t1[0]*t1[0]+t1[1]*t1[1]));
+            t1[0] = normal[1]/norm;
+            t1[1] = -normal[0]/norm;
+
+            t2[0] = -normal[0]*normal[2]/norm;
+            t2[1] = -normal[1]*normal[2]/norm;
+            t2[2] = normal[0]*normal[0]+normal[1]*normal[1]/norm;
+        } else{
+            t1[0] = 1.0;
+            t2[1] = 1.0;
+        }
+
+        T_k1(0,0) = normal[0]*t1[0];
+        T_k1(0,1) = normal[1]*t1[1];
+        T_k1(0,2) = normal[2]*t1[2];
+        T_k1(0,3) = normal[1]*t1[2]+normal[2]*t1[1];
+        T_k1(0,4) = normal[2]*t1[0]+normal[0]*t1[2];
+        T_k1(0,5) = normal[0]*t1[1]+normal[1]*t1[0];
+        
+        T_k2(0,0) = normal[0]*t2[0];
+        T_k2(0,1) = normal[1]*t2[1];
+        T_k2(0,2) = normal[2]*t2[2];
+        T_k2(0,3) = normal[1]*t2[2]+normal[2]*t2[1];
+        T_k2(0,4) = normal[2]*t2[0]+normal[0]*t2[2];
+        T_k2(0,5) = normal[0]*t2[1]+normal[1]*t2[0];
+    }
+    
+    noalias(A1) = prod(trans(p_k),trans(N_k));
+    noalias(A2) = prod(N_k,p_k);
+    noalias(A) += mPenaltyNormal * prod(A1, A2);
+
+    noalias(A1) = prod(trans(p_k),trans(T_k1));
+    noalias(A2) = prod(T_k1,p_k);
+
+    // Finally we add the contributiosn to our local system (A, b)
+    noalias(A) += mPenaltyTangent*prod(A1, A2);
+    noalias(b) += mPenaltyNormal*prod(trans(p_k),trans(N_k)) * itNode->GetValue(CONTACT_PRESSURE);
+
+    //PART 2: Contributions from contact nodes: regard all nodes from the patch which are in contact
+    // Patch center node:
+    if (itPatchNode->Has(CONTACT_PRESSURE)){
```

Comment:
you already assumed in the line before that the node has CONTACT_PRESSURE

Reply:
btw, this function is too long. I really think it would be a better solution to have a 

SPRErrorProcess (or whatever name) and to derive from it a SPRContactErrorProcess

Reply:
Right now the detection is automatic, I don't like the idea to separate the process

Reply:
as i am telling detection is automatic but it does have a performance impact...also if you don't separate than you need to test separately with the behaviour in the two different cases, with and without contact